### ISSUE TITLE
Core/Gameobject: Gossip Hello executed twice

### ISSUE THREAD

**Issue by Killyana** [TrinityCore/TrinityCore#19819]
**Description:** When a gob is opened the OnGossipHello is executed once one cliched and once when the gob is opened.

**Current behaviour:** OnGossipHello is executed twice

**Expected behaviour:** OnGossipHello must be executed only once if there's an opening action.

**Steps to reproduce the problem:**

1. .go c id 17364
2. .quest add 9544
3. .add 23801 
4. .Spam click a cage
5. .notice the OnGossipHello is executed before the opening of the gob and it could also be executed at the opening depending on the script.

**Branch(es):** 3.3.5

**TC rev. hash/commit:**  rev. f2b0819e5368

**Comment by Keader** [comment]
True, i have the same issue in Razorscale
Using this script: https://gist.github.com/Keader/c89da63e9d67259ca85c663765d7aaef

**Comment by ariel-** [comment]
In fact is executed thrice:

- GameObject::OnUse
- WorldSession::HandleGameobjectReportUse (called with second param true)
- Spell::SendLoot

What I propose is streamline GossipHello just like the Creature counterpart, with only parameter player, and add other hooks like OnLoot and OnReportUse.

It'll break more custom scripts probably, but in the end it'll be cleaner and easier to script GameObjects

**Comment by Shauren** [comment]
Look at GameObjectAI gossip hello, that one has an extra arg to determine which opcode triggered the action

**Comment by ariel-** [comment]
There is no current way of distinguishing between GameObject::OnUse and Spell::SendLoot though (in both cases the extra arg is false)

**Comment by joshwhedon** [comment]
Personnally I dislike the idea of breaking more scripts by changing an API again, but if it is the right way to do it, well ... Good thing for the long term.
But in the end, why is GossipHello called multiple times now and wasn't before ? I thought all that was done was reorganising the AIs and streamlining them.

**Comment by ariel-** [comment]
> But in the end, why is GossipHello called multiple times now and wasn't before ? 

It was before too

**Comment by tkrokli** [comment]
Is this issue related to what happens in master with the quest gossip, where the gossip moves on the the [Continue] page before player is able to read the quest? (This happens when quest giver has got a direct follow-up quest to the one you hand in.)

**Comment by joshwhedon** [comment]
@ariel- I don't think so no. A lot of problems started happening since recently (example, when you click on the orb in Utgarde Pinnacle, it activates 2 of the bosses instead of one). So before something must have been keeping the gossip hello from triggering twice.

**Comment by ariel-** [comment]
The UP script was rewritten recently (#18712), don't confuse it.

**Comment by joshwhedon** [comment]
You say recently, it was rewritten in January. And that issue was reported to me only after merging the gossip changes.
The reason is simple actually. Before the gossip changes, the gossipHello in HandleGameobjectReportUse was only called for SmartAI, and the one in Use was only called for scripts. Following the rewamp, they have been merged into a single one, and therefore the method is called twice. I think one of the two calls should be removed.

**Comment by ariel-** [comment]
- The GossipHello on HandleGameObjectReportUse wasn't changed, only called AI* version and is called for every AI, not only SmartAI.

- GameObject::Use one called both (script and AI)

- Spell::SendLoot called both, too.

Those discrepancies were the reason I unified APIs.

**Comment by joshwhedon** [comment]
I know, and I'm not saying unifying it is wrong. I'm just saying that before, HandleGameobjectReportUse was called only for SmartAI, Use was called only for GameObjectScript, and now, HandleGameobjectReportUse  AND Use are called for both SmartAI and GameObjectScript, due to the unitification, which means that following the change, Palehoof orb for example, has its GossipHello called twice, while before, it was only called once.
Also, why do we need to call GossipHello in both HandleGameobjectReportUse  and Use ? is there a situation where only one is called and not the other ?

**Comment by ariel-** [comment]
Overriding GossipHello for report use case allows it to skip achievement criteria tracking (ACHIEVEMENT_CRITERIA_TYPE_USE_GAMEOBJECT) if you return true from that function.

Be sure to not be confusing "AI" with "SmartAI", the GO in the description (Bristlelimb Cage) is scripted in C++, not SmartAI

**Comment by Seebaby** [comment]
Heyho how would the one fix look for it is still a novice in c ++

**Comment by tkrokli** [comment]
@Seebaby : ask in TC Forum pages, **https://community.trinitycore.org/** or ask in [**IRC: #trinity @ irc.rizon.net** (Rules)](https://trinitycore.atlassian.net/wiki/display/tc/IRC).

The TrinityCore issue tracker is **not** a forum.
